### PR TITLE
ipn: use tstime

### PR DESCRIPTION
**Pull request by clairew**
Updates #8587
Signed-off-by: Claire Wang <claire@tailscale.com>

### CODE REVIEW

File: ipn/ipnlocal/local.go [hunk 1]
```diff
@@ -311,6 +313,7 @@ func NewLocalBackend(logf logger.Logf, logID logid.PublicID, sys *tsd.System, lo
 		em:             newExpiryManager(logf),
 		gotPortPollRes: make(chan struct{}),
 		loginFlags:     loginFlags,
+		clock:          &tstime.StdClock{},
```
Comment (sailorfrag) [inline ipn/ipnlocal/local.go:316]:
```suggestion
		clock:          tstime.StdClock{},
```

File: ipn/localapi/localapi.go [hunk 2]
```diff
@@ -1106,7 +1108,7 @@ func (h *Handler) serveFiles(w http.ResponseWriter, r *http.Request) {
 				http.Error(w, "invalid waitsec", http.StatusBadRequest)
 				return
 			}
-			deadline := time.Now().Add(time.Duration(d) * time.Second)
+			deadline := h.clock.Now().Add(time.Duration(d) * time.Second)
 			var cancel context.CancelFunc
 			ctx, cancel = context.WithDeadline(ctx, deadline)
```
Comment (sailorfrag) [inline ipn/localapi/localapi.go:1113]: This'll be slightly awkward until `tstime.Clock` also has a way to implement `context.WithDeadline` against simulated time.

**Review by sailorfrag — APPROVED** [review]

**Comment by bradfitz** [comment]
> [sailorfrag](https://github.com/sailorfrag) approved these changes [17 hours ago](https://github.com/tailscale/tailscale/pull/8597#pullrequestreview-1527211711)

Let's not approve draft PRs. If a PR is ready to be reviewed, it should be un-drafted.

**Comment by sailorfrag** [comment]
> > [sailorfrag](https://github.com/sailorfrag) approved these changes [17 hours ago](https://github.com/tailscale/tailscale/pull/8597#pullrequestreview-1527211711)
> 
> Let's not approve draft PRs. If a PR is ready to be reviewed, it should be un-drafted.

Oh fair. This one is straightforward enough that I think it's fine to un-draft.

File: ipn/localapi/localapi.go (hunk 2 above)
Comment (clairew) [reply ipn/localapi/localapi.go:1113]: added followup issue here https://github.com/tailscale/tailscale/issues/8604

File: ipn/ipnlocal/expiry_test.go [hunk 3]
```diff
@@ -254,7 +254,7 @@ func TestNextPeerExpiry(t *testing.T) {
 	t.Run("ClockSkew", func(t *testing.T) {
 		t.Logf("local time:  %q", now.Format(time.RFC3339))
 		em := newExpiryManager(t.Logf)
-		em.timeNow = func() time.Time { return now }
+		em.clock = tstest.NewClock(tstest.ClockOpts{Start: time.Unix(1675725516, 0)})
```
Comment (bradfitz) [inline ipn/ipnlocal/expiry_test.go:257]: where these three numbers coming from? What's with the `now` variable that was previously being used that's not?

File: ipn/ipnlocal/local.go [hunk 4]
```diff
@@ -1379,13 +1383,13 @@ func (b *LocalBackend) Start(opts ipn.Options) error {
 			// prevent it from restarting our map poll
 			// HTTP request (via doSetHostinfoFilterServices >
 			// cli.SetHostinfo). In practice this is very quick.
-			t0 := time.Now()
-			timer := time.NewTimer(time.Second)
+			t0 := b.clock.Now()
+			timer, timerChannel := b.clock.NewTimer(time.Second)
 			select {
 			case <-b.gotPortPollRes:
 				b.logf("[v1] got initial portlist info in %v", time.Since(t0).Round(time.Millisecond))
```
Comment (bradfitz) [inline ipn/ipnlocal/local.go:1390]: time.Since here is mixing clocks

File: ipn/ipnlocal/local.go [hunk 5]
```diff
@@ -360,7 +364,7 @@ func NewLocalBackend(logf logger.Logf, logID logid.PublicID, sys *tsd.System, lo
 
 type componentLogState struct {
 	until time.Time
-	timer *time.Timer // if non-nil, the AfterFunc to disable it
+	timer tstime.TimerController // if non-nil, the AfterFunc to disable it
```
Comment (bradfitz) [inline ipn/ipnlocal/local.go:367]: comment is no longer right?

Comment (clairew) [reply ipn/ipnlocal/local.go:367]: No still is the same - tstime.TimerController wraps time.Timer behavior

File: ipn/ipnlocal/local.go (hunk 4 above)
Comment (clairew) [reply ipn/ipnlocal/local.go]: replaced w/ `clock.Now().Sub(t0)`

Comment (bradfitz) [reply ipn/ipnlocal/local.go]: can we add a `Since` method instead to keep the code looking the same?

File: ipn/ipnlocal/local.go [hunk 6]
```diff
@@ -2397,7 +2401,7 @@ func (b *LocalBackend) StartLoginInteractive() {
 
 func (b *LocalBackend) Ping(ctx context.Context, ip netip.Addr, pingType tailcfg.PingType) (*ipnstate.PingResult, error) {
 	if pingType == tailcfg.PingPeerAPI {
-		t0 := time.Now()
+		t0 := b.clock.Now()
```
Comment (sailorfrag) [inline ipn/ipnlocal/local.go:2404]: This has a corresponding `time.Since` that needs updating.

**Review by sailorfrag — APPROVED** [review]

**Review by sailorfrag — APPROVED** [review]

**Pull request merged**